### PR TITLE
Add error for different use case

### PR DESCRIPTION
**Pull request by gwendalF**
Define custom errors type. More type could be use if needed and remove the anyhow completely. Anyhow is better for application developper who want to manage every errors. For a library it's better to provide different errors and let the consumer decide what to do
https://github.com/ARK-Builders/arklib/issues/43

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+use thiserror::Error;
+
+pub type Result<T> = std::result::Result<T, ArklibError>;
+
+#[derive(Error, Debug)]
+pub enum ArklibError {
+    #[error("IO error")]
+    Io(#[from] std::io::Error),
+    #[error("Invalid path for ressource")]
```
Comment (kirillt) [inline src/errors.rs:11]: Resource with one `s` :wink:  I already fixed that once, so now telling in comments..

File: src/index.rs [hunk 2]
```diff
@@ -90,14 +89,16 @@ impl ResourceIndex {
...
+                            str.parse().map_err(|_| ArklibError::Parse)?,
+                        ))
                     };
 
                     let id: ResourceId = {
-                        let str = parts.next().ok_or(Error::msg(
+                        let str = parts.next().ok_or(anyhow::Error::msg(
                             "Couldn't find second part of index entry",
                         ))?;
                         ResourceId::from_str(str)?
```
Comment (kirillt) [inline src/index.rs:104]: But why `anyhow` here?

File: src/index.rs [hunk 3]
```diff
@@ -386,7 +381,9 @@ impl ResourceIndex {
                             }
 
                             // the caller must have ensured that the path was updated
-                            return Err(Error::msg("Nothing to update"));
+                            return Err(ArklibError::Collision(
+                                "Nothing to update".into(),
```
Comment (kirillt) [inline src/index.rs:385]: Could you also change the message to `New content has the same id.`

File: src/index.rs [hunk 4]
```diff
@@ -515,17 +514,14 @@ fn discover_paths<P: AsRef<Path>>(
...
     if metadata.is_dir() {
-        return Err(Error::msg("DirEntry is directory"));
+        return Err(ArklibError::Path);
     }
 
     let size = metadata.len();
     if size == 0 {
-        return Err(Error::msg("Empty resource"));
+        return Err(ArklibError::Path);
     }
```
Comment (kirillt) [inline src/index.rs:525]: Looks like a problem with the content, not with the path.

File: src/index.rs [hunk 5]
```diff
@@ -515,17 +514,14 @@ fn discover_paths<P: AsRef<Path>>(
...
 }
 
-fn scan_entry(
-    path: &CanonicalPath,
-    metadata: Metadata,
-) -> Result<IndexEntry, Error> {
+fn scan_entry(path: &CanonicalPath, metadata: Metadata) -> Result<IndexEntry> {
     if metadata.is_dir() {
-        return Err(Error::msg("DirEntry is directory"));
+        return Err(ArklibError::Path);
```
Comment (kirillt) [inline src/index.rs:519]: Detailed message would be good unless that's the only `ArklibError::Path`  in the whole codebase.

File: src/index.rs (hunk 2 above)
Comment (gwendalF) [reply src/index.rs]: I fix it using Parse Error seems it's more like a parsing issue

File: src/index.rs (hunk 3 above)
Comment (gwendalF) [reply src/index.rs:380]: I updated the message

File: src/index.rs (hunk 4 above)
Comment (gwendalF) [reply src/index.rs]: I used std::io::Error::new(std::io::Errorkind::InvalidData

File: src/index.rs (hunk 5 above)
Comment (gwendalF) [reply src/index.rs]: I have added a String message on the Path error

**Review by kirillt — APPROVED** [review]

**Pull request merged**